CMakeLists.txt

```diff
@@ -85,14 +85,46 @@ elseif(KernelArchARM)
     endif()
 
 elseif(KernelArchRiscV)
+    # Determine if GNU toolchain is used and if yes, whether binutils version >= 2.38
+    if(CMAKE_ASM_COMPILER_ID STREQUAL "GNU")
+        # Not using CMAKE_ASM_COMPILER directly because it is set to ${CROSS_COMPILER_PREFIX}gcc in gcc.cmake
+        # => can't change that easily without messing up the build system
+        execute_process(COMMAND ${CROSS_COMPILER_PREFIX}as --version OUTPUT_VARIABLE _gas_version)
+        # Extract version number
+        string(
+            REGEX
+                MATCH
+                "([0-9]+\.)?([0-9]+\.)?[0-9]+"
+                _gas_version_num
+                "${_gas_version}"
+        )
+        # Check whether binutils >= 2.38
+        if(_gas_version_num VERSION_GREATER_EQUAL "2.38")
+            # Require gcc >= 10, otherwise it gets confused by the RISC-V Z extensions
+            if(CMAKE_C_COMPILER_VERSION VERSION_LESS "10")
+                message(
+                    FATAL_ERROR
+                        "Incompatible GCC and GNU binutils versions ${CMAKE_C_COMPILER_VERSION} and ${_gas_version_num}"
+                )
+            else()
```

Comment:
As GCC 11.3 requires adding the extensions by default and would seem that using GCC 11.3 is required for using binutils >=2.38, could this compatibility check be reduced to just checking GCC version:
```
        if(
            (CMAKE_C_COMPILER_ID STREQUAL "GNU")
            AND (CMAKE_C_COMPILER_VERSION VERSION_GREATER_EQUAL "11.3.0")
        )
                string(APPEND KernelRiscVMachFeatureModifiers "_zicsr_zifencei")
        endif()
```

Reply:
@kent-mcleod @axel-h I'm just replying here once since your comments about GCC versions seem to point into the same direction, I hope you don't mind! :)

As far as I understood the GCC and binutils changelogs correctly, it should technically be sufficient to check for GCC >= 11.3 because that's when it by default starts to ship with binutils 2.38. However, this combination cannot be guaranteed. In my local tests, I compiled multiple versions of GCC and the GNU binutils and checked their compatibility with regard to the extension flags. This is where my checks originate from: specifying the Zicsr and Zifenci extensions is only necessary from binutils 2.38 onwards but GCC already supports parsing/forwarding those flags to the assembler backend from version 10 onwards.
Even though it is unlikely that a combination of GCC 9.x and binutils 2.38 is encountered "in the wild", my checks explicitly make sure that the project configuration already fails and we don't encounter unexpected errors during the build process.

I agree that my checks are quite verbose, I just wanted to rule out as many error causes as possible. Do you think that a best-effort approach just checking the GCC version here is sufficient?

Reply:
What exactly happens when you use binutils 2.38 with an older GCC, do RISC-V builds raise an error due to some incompatibility eventually? What is the error? This seems be a general RISC-V issue then, that is no just seL4 specific?

Reply:
Yes, this is a general RISC-V toolchain issue. E.g., GCC 9.1 in combination with binutils 2.38 errors out with `unsupported ISA subset 'z'`, it seems as if it simply cannot parse the ISA correctly and pass it on to the assembler.
Since this is not an officially supported combination, it makes sense that there are breaking changes/incompatibilities at some point.

Reply:
There are the "official" seL4 docker containers that contain the built environment that are supported and that are used for the verification. Saying that all other environment are "don't care" is a bit harsh, but adding too many check for tool combinations would also become overkill. Given this is a hard break in the whole GCC/RISC-V ecosystem, but it also yields an error message `unsupported ISA subset 'z` that one's favorite search engine will provide enough hints in the long run (e.g. https://www.mail-archive.com/ubuntu-bugs@lists.ubuntu.com/msg5993913.html), I think limiting this to checking GCC 11.3 and adding the extenstion "flags" is good enough. In the near future this might be a bit annoying, but in the long run I'd expect this to resolve itself. 

While we are at that topic, it seems the LLVM/clang world is totally unaffected of this?

Reply:
I'm sorry for the late reply, I was (and unfortunately still am) quite busy this week.

Just checking for the GCC version works for me, it makes the checks way easier and shorter.
I'll rework and force-push to my branch.